pkg/api/transport_custom.go

```diff
@@ -0,0 +1,223 @@
+type CustomTransport struct {
+	rt http.RoundTripper
+
+	// UserAgent settings
+	agent string
+
+	// Retry settings
+	retries int
+	backoff time.Duration
+
+	// Verbose settings
+	verbose    bool
+	count      int64
+	redactAuth bool
+	writer     io.Writer
+}
+
+// CustomTransportCfg is used to configure a CustomTransport.
+type CustomTransportCfg struct {
+	// RoundTripper to use for http calls.
+	RoundTripper http.RoundTripper
+
+	// UserAgent if specified, it sets the user agent on all outgoing requests.
+	UserAgent string
+
+	// Number of retries to perform on request timeout.
+	Retries int
+
+	// Cooldown time between retried requests.
+	Backoff time.Duration
+
+	// Verbose settings
+	Verbose    bool
+	RedactAuth bool
+	Writer     io.Writer
+}
+
+func (cfg CustomTransportCfg) validate() error {
+	var merr = multierror.NewPrefixed("invalid custom transport settings")
+	if cfg.RoundTripper == nil {
+		merr = merr.Append(errors.New("roundtripper cannot be nil"))
+	}
+
+	if cfg.Verbose && cfg.Writer == nil {
+		merr = merr.Append(errors.New("verbose set to true, but no writer has been set"))
+	}
+
+	return merr.ErrorOrNil()
+}
+
+// NewCustomTransport creates a new CustomTransport. See the structure's Goddoc
+// to learn about what it does.
+func NewCustomTransport(cfg CustomTransportCfg) (*CustomTransport, error) {
+	if err := cfg.validate(); err != nil {
+		return nil, err
+	}
+
+	if cfg.UserAgent == "" {
+		cfg.UserAgent = DefaultUserAgent
+	}
+
+	if cfg.Backoff.Microseconds() <= 0 {
+		cfg.Backoff = defaultBackoff
+	}
+
+	return &CustomTransport{
+		rt:         cfg.RoundTripper,
+		agent:      cfg.UserAgent,
+		retries:    cfg.Retries,
+		backoff:    cfg.Backoff,
+		verbose:    cfg.Verbose,
+		redactAuth: cfg.RedactAuth,
+		writer:     cfg.Writer,
+	}, nil
+}
+
+// RoundTrip wraps http.DefaultTransport.RoundTrip to keep track
+// of the current request.
+func (t *CustomTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	// UserAgent header handling
+	req.Header.Set(userAgentHeader, t.agent)
+
+	return t.doRoundTrip(req)
+}
+
+// doRoundTrip performs an http call with the specified request and retries the
+// request when the returned error is context.DeadlineExceeded (timeout).
+func (t *CustomTransport) doRoundTrip(req *http.Request) (res *http.Response, err error) {
+	for i := -1; i < t.retries; i++ {
+		count := atomic.AddInt64(&t.count, 1)
+		res, err = t.rt.RoundTrip(req)
+
+		if t.verbose {
+			handleVerboseRequest(t.writer, req, count, t.redactAuth)
+			if res != nil {
+				handleVerboseResponse(t.writer, res, count)
+			}
+		}
+
+		// Return early when err is empty or not a timeout.
+		if err == nil || !errors.Is(err, context.DeadlineExceeded) {
+			return
+		}
+
+		if t.verbose {
+			var msg = "request %d/%d timed out, giving up.\n"
+			if i+2 <= t.retries {
+				msg = "request %d/%d timed out, retrying...\n"
+			}
+			_, _ = fmt.Fprintf(t.writer, msg, count, t.retries+1)
```

Comment:
`_, _` ? Is this a linter thing?

Reply:
It is, I think it's not required here, actually. I'll remove this one.